### ISSUE TITLE
Keras mask propagation between recurrent, pooling and Dense layers from Masking layer

### ISSUE THREAD

**Issue by ugolbck** [tensorflow/tensorflow#49262]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Debian 10
- TensorFlow installed from (source or binary): pip
- TensorFlow version (use command below): 2.4.1
- Python version: 3.8

**Current behavior**

From what I understand, the mask created by the Masking layer propagates only through certain layers, and I am having trouble figuring out through which layers and if the end result will vary depending on this mask's absence.

In my example, I have inputs of shape `(None, timesteps, features)` already embedded and 0-padded in order to have constant `timesteps` value. I want to pass the inputs through some recurrent layers, and then perform either classification (or token classification, but the problem remains the same). The questions are:
1) Do I have a problem if my output layer does not have an `input_mask` (0-padding computed in loss, for example)?
2) Do I need to make custom layers every time I need my mask to propagate (when not modifying the `timesteps` dimension)? 

For each of the four minimal examples below, I run this piece of code to figure out which layer has or hasn't an `input_mask`  attribute:
```
model = make_model(120, 768)
print(model.summary())
for layer in model.layers:
    print(layer.name, layer.input_mask)
``` 

**Example 1: Token classification with Masking+LSTM+Dense**
```
def make_model(m_len: int, emb_dim: int):
    inputs = Input(shape=(m_len, emb_dim))
    x = Masking(mask_value=0)(inputs)
    x = Bidirectional(LSTM(256, return_sequences=True))(x)
    output = Dense(1, activation="sigmoid")(x)
    mod = Model(inputs, output)
    mod.compile(
        optimizer="Adam",
        loss="binary_crossentropy",
        metrics=["accuracy"]
    )
    return mod
```
output:
```
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
input_1 (InputLayer)         [(None, 120, 768)]        0         
_________________________________________________________________
masking (Masking)            (None, 120, 768)          0         
_________________________________________________________________
bidirectional (Bidirectional (None, 120, 512)          2099200   
_________________________________________________________________
dense (Dense)                (None, 120, 1)            513       
=================================================================
Total params: 2,099,713
Trainable params: 2,099,713
Non-trainable params: 0
_________________________________________________________________
None
input_1 None
masking None
bidirectional KerasTensor(type_spec=TensorSpec(shape=(None, 120), dtype=tf.bool, name=None), name='masking/Squeeze:0')
dense KerasTensor(type_spec=TensorSpec(shape=(None, 120), dtype=tf.bool, name=None), name='Placeholder_2:0')
```
The mask is clearly propagated through both the recurrent layer and is also here in the Dense layer since the `timesteps` dimension hasn't changed.

**Example 2: Classification with Masking+LSTM+Dense**

Same code as above but setting `return_sequences=False`. output:

```
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
input_1 (InputLayer)         [(None, 120, 768)]        0         
_________________________________________________________________
masking (Masking)            (None, 120, 768)          0         
_________________________________________________________________
bidirectional (Bidirectional (None, 512)               2099200   
_________________________________________________________________
dense (Dense)                (None, 1)                 513       
=================================================================
Total params: 2,099,713
Trainable params: 2,099,713
Non-trainable params: 0
_________________________________________________________________
None
input_1 None
masking None
bidirectional KerasTensor(type_spec=TensorSpec(shape=(None, 120), dtype=tf.bool, name=None), name='masking/Squeeze:0')
dense None
```
Works as expected so far, the dimension reduction doesn't allow a mask to pass on.

**Example 3: Classification with Masking+LSTM**
```
def make_model(m_len: int, emb_dim: int):
    inputs = Input(shape=(m_len, emb_dim))
    x = Masking(mask_value=0)(inputs)
    x = Bidirectional(LSTM(256, return_sequences=True))(x)
    output = LSTM(1, activation="sigmoid")(x)
    mod = Model(inputs, output)
    mod.compile(
        optimizer="Adam",
        loss="binary_crossentropy",
        metrics=["accuracy"]
    )
    return mod
```
output
```
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
input_1 (InputLayer)         [(None, 120, 768)]        0         
_________________________________________________________________
masking (Masking)            (None, 120, 768)          0         
_________________________________________________________________
bidirectional (Bidirectional (None, 120, 512)          2099200   
_________________________________________________________________
lstm_1 (LSTM)                (None, 1)                 2056      
=================================================================
Total params: 2,101,256
Trainable params: 2,101,256
Non-trainable params: 0
_________________________________________________________________
None
input_1 None
masking None
bidirectional KerasTensor(type_spec=TensorSpec(shape=(None, 120), dtype=tf.bool, name=None), name='masking/Squeeze:0')
lstm_1 KerasTensor(type_spec=TensorSpec(shape=(None, 120), dtype=tf.bool, name=None), name='Placeholder_2:0')
```
Even though the `timesteps` dim is removed, the mask is still effective on the output layer, is this behavior expected?

**Example 4: Classification with Masking+LSTM+AveragePooling+Flatten+Dense**
```
def make_model(m_len: int, emb_dim: int):
    inputs = Input(shape=(m_len, emb_dim))
    x = Masking(mask_value=0)(inputs)
    x = Bidirectional(LSTM(256, return_sequences=True))(x)
    x = AveragePooling1D(pool_size=512, data_format='channels_first')(x)
    x = Flatten()(x)
    output = Dense(1, activation="sigmoid")(x)
    mod = Model(inputs, output)
    mod.compile(
        optimizer="Adam",
        loss="binary_crossentropy",
        metrics=["accuracy"]
    )
    return mod
```
output:
```
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
input_1 (InputLayer)         [(None, 120, 768)]        0         
_________________________________________________________________
masking (Masking)            (None, 120, 768)          0         
_________________________________________________________________
bidirectional (Bidirectional (None, 120, 512)          2099200   
_________________________________________________________________
average_pooling1d (AveragePo (None, 120, 1)            0         
_________________________________________________________________
flatten (Flatten)            (None, 120)               0         
_________________________________________________________________
dense (Dense)                (None, 1)                 121       
=================================================================
Total params: 2,099,321
Trainable params: 2,099,321
Non-trainable params: 0
_________________________________________________________________
None
input_1 None
masking None
bidirectional KerasTensor(type_spec=TensorSpec(shape=(None, 120), dtype=tf.bool, name=None), name='masking/Squeeze:0')
average_pooling1d KerasTensor(type_spec=TensorSpec(shape=(None, 120), dtype=tf.bool, name=None), name='Placeholder_2:0')
flatten None
dense None
```
This one I am a bit less confident about, but averaging on the last dimension doesn't change the `timesteps` dimension and yet the mask isn't propagated. I thought maybe the averaging could create "0" vectors that would be ignored afterwards but the mask is computed by Masking and is 120 long boolean here, no matter what. Is there a reason why AveragePooling1D is not passing the mask to the Flatten layer?

**Comment by qlzh727** [comment]
Sorry for the late reply.

For example 3, the result is expected (since you are printing layer.input_mask). For the last LSTM layer, its input has a shape of (None, 120, 512), which contains the timestep dim. The LSTM layer need to properly ignore those masked inputs. The output of the LSTM has no timestep dim, so the mask propagation stops there.

For example 4, the AveragePooling1D layer doesn't support mask at all (even its input is 3D). If you check the code, the call method doesn't take a kwarg mask, which indicate it ignores the masks. https://github.com/keras-team/keras/blob/a5a6a53eceb4bb0957fcbe577f56941ae8062d8f/keras/layers/pooling.py#L65

I am not sure if this is bug or a feature. Also, propagating the mask to next layer seems problematic if the value of pooling/strides is not 1.

**Comment by qlzh727** [comment]
Closing as working as intended. Feel free to reopen this issue if there is any other items need to be addressed.
